### DISCUSSION TITLE
[Question] When I draw sub-structure with aromatic atom by "Draw.MolDraw2DCairo", I get error ""AtomKekulizeException

### DISCUSSION THREAD

**Discussion by mohammad-saber** [rdkit/rdkit#4418] [Q&A]
Hi, Thank you for your time to read my question. 
I am trying to draw sub-structures of a molecule containing aromatic ring by "Draw.MolDraw2DCairo" and save it as PNG file. Following is the environmnent and code:

**Environment:**
- RDKit Version: '2020.09.1'
- Operating system: masOS 11.5
- Python version: 3.6

**Code:**

```
from rdkit import Chem
from rdkit.Chem import Draw
from rdkit.Chem.Draw import rdMolDraw2D
from IPython.display import SVG
from rdkit.Chem import rdDepictor

smiles_parent = 'O=C(CC1=CC=C(OC(C(CCCCCCCCCCCCCCCC)CCCCCCCCCCCCCCCC)=O)C=C1)OCCSSCCOC(NCCCCN(C)C)=O'
mol_parent = Chem.MolFromSmiles(smiles_parent)

sub_structures = ['N(C)(C)CC', 'c(cc)(cc)CC']
legends = sub_structures

# Read sub_structures
sub_mols = []
for smiles in sub_structures:
    if any(True for c in smiles if c.islower()):   # If there is any aromatic atom
        mol = Chem.MolFromSmiles(smiles, sanitize=False) 
        Chem.SanitizeMol(mol, sanitizeOps=(Chem.SanitizeFlags.SANITIZE_ALL^Chem.SanitizeFlags.SANITIZE_KEKULIZE^Chem.SanitizeFlags.SANITIZE_SETAROMATICITY))
        sub_mols.append(mol)
    else:
        mol = Chem.MolFromSmiles(smiles)  
        sub_mols.append(mol)


# Drawing
fullSize = (400, 200)   # size of image
subImgSize = (200, 200)
rdDepictor.Compute2DCoords(mol_parent)   

d2d = Draw.MolDraw2DCairo(fullSize[0], fullSize[1], subImgSize[0], subImgSize[1])
d2d.drawOptions().prepareMolsBeforeDrawing=False   # not to kekulize the bonds before drawing
d2d.DrawMolecules(sub_mols, legends=legends)
d2d.FinishDrawing()
image = d2d.GetDrawingText()   

with open('image.png','wb+') as outfile:
    outfile.write(image)
```

But, I get the error:

`AtomKekulizeException: non-ring atom 0 marked aromatic
`

It seems that `d2d.drawOptions().prepareMolsBeforeDrawing=False` doesn't work here.

I can use "MolDraw2DSVG" and save as SVG file by the following code:

```
d2d = rdMolDraw2D.MolDraw2DSVG(fullSize[0], fullSize[1], subImgSize[0], subImgSize[1])
d2d.drawOptions().prepareMolsBeforeDrawing=False   # not to kekulize the bonds before drawing
d2d.DrawMolecules(sub_mols, legends=legends)   
d2d.FinishDrawing()
image = SVG(d2d.GetDrawingText())   

with open('image.svg','wb+') as outfile:
    outfile.write(image.data.encode("utf-8"))
```

However, I prefer to save drawings as PNG file. If you have any idea to solve this issue, please let me know.

**Comment by DavidACosgrove** [comment]
Hi,
I believe this problem has been fixed in later versions of the toolkit. I
suggest you upgrade to the latest version. If the problem persists, please
re-post.
Regards,
Dave


On Sat, 14 Aug 2021 at 03:42, mohammad-saber ***@***.***>
wrote:

> Hi, Thank you for your time to read my question.
> I am trying to draw sub-structures of a molecule containing aromatic ring
> by "Draw.MolDraw2DCairo" and save it as PNG file. Following is the
> environmnent and code:
>
> *Environment:*
>
>    - RDKit Version: '2020.09.1'
>    - Operating system: masOS 11.5
>    - Python version: 3.6
>
> *Code:*
>
> from rdkit import Chem
> from rdkit.Chem import Draw
> from rdkit.Chem.Draw import rdMolDraw2D
> from IPython.display import SVG
> from rdkit.Chem import rdDepictor
>
> smiles_parent = 'O=C(CC1=CC=C(OC(C(CCCCCCCCCCCCCCCC)CCCCCCCCCCCCCCCC)=O)C=C1)OCCSSCCOC(NCCCCN(C)C)=O'
> mol_parent = Chem.MolFromSmiles(smiles_parent)
>
> sub_structures = ['N(C)(C)CC', 'c(cc)(cc)CC']
> legends = sub_structures
>
> # Read sub_structures
> sub_mols = []
> for smiles in sub_structures:
>     if any(True for c in smiles if c.islower()):   # If there is any aromatic atom
>         mol = Chem.MolFromSmiles(smiles, sanitize=False)
>         Chem.SanitizeMol(mol, sanitizeOps=(Chem.SanitizeFlags.SANITIZE_ALL^Chem.SanitizeFlags.SANITIZE_KEKULIZE^Chem.SanitizeFlags.SANITIZE_SETAROMATICITY))
>         sub_mols.append(mol)
>     else:
>         mol = Chem.MolFromSmiles(smiles)
>         sub_mols.append(mol)
>
>
> # Drawing
> fullSize = (400, 200)   # size of image
> subImgSize = (200, 200)
> rdDepictor.Compute2DCoords(mol_parent)
>
> d2d = Draw.MolDraw2DCairo(fullSize[0], fullSize[1], subImgSize[0], subImgSize[1])
> d2d.drawOptions().prepareMolsBeforeDrawing=False   # not to kekulize the bonds before drawing
> d2d.DrawMolecules(sub_mols, legends=legends)
> d2d.FinishDrawing()
> image = d2d.GetDrawingText()
>
> with open('image.png','wb+') as outfile:
>     outfile.write(image)
>
> But, I get the error:
>
> AtomKekulizeException: non-ring atom 0 marked aromatic
>
> It seems that d2d.drawOptions().prepareMolsBeforeDrawing=False doesn't
> work here.
>
> I can use "MolDraw2DSVG" and save as SVG file by the following code:
>
> d2d = rdMolDraw2D.MolDraw2DSVG(fullSize[0], fullSize[1], subImgSize[0], subImgSize[1])
> d2d.drawOptions().prepareMolsBeforeDrawing=False   # not to kekulize the bonds before drawing
> d2d.DrawMolecules(sub_mols, legends=legends)
> d2d.FinishDrawing()
> image = SVG(d2d.GetDrawingText())
>
> with open('image.svg','wb+') as outfile:
>     outfile.write(image.data.encode("utf-8"))
>
> However, I prefer to save drawings as PNG file. If you have any idea to
> solve this issue, please let me know.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4418>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FVY7QCWFRXSI4XY3FDT4XJXTANCNFSM5CETN7MA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>
-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Reply by mohammad-saber** [reply]
Thank you for your help. 

In order to upgrade, should I uninstall RDKit and install it again? 

I installed RDKit by `conda install -c conda-forge rdkit` in a conda virtual environment. Could you let me know how to uninstall it?

Also, which of the following commands will install the latest version:

```
conda install -c conda-forge rdkit
conda install -c rdkit rdkit
```

Thank you for your help.

**Comment by DavidACosgrove** [comment]
I always find it easiest to install the new version into a separate conda
environment and then when I'm happy it's all working I can delete the old
one.  So for this I would do something like

conda create -c conda-forge -n RDKit.2021.03.4 rdkit

This makes a new conda environment called RDKit.2021.03.4 which you
then activate:

conda activate RDKit.2021.03.4

The conda documentation gives much information about managing
environments and is well worth consulting.


On Sat, Aug 14, 2021 at 4:20 PM mohammad-saber ***@***.***>
wrote:

> Thank you for your help.
>
> In order to upgrade, should I uninstall RDKit and install it again?
>
> I installed RDKit by conda install -c conda-forge rdkit in a conda
> virtual environment. Could you let me know how to uninstall it?
>
> Also, which of the following commands will install the latest version:
>
> conda install -c conda-forge rdkit
> conda install -c rdkit rdkit
>
> Thank you for your help.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4418#discussioncomment-1185001>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FUQHV73IYP5JBAVBMDT42CUPANCNFSM5CETN7MA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk
